src/main/scala/org/allenai/plugins/DeployPlugin.scala

```diff
@@ -221,56 +228,79 @@ object DeployPlugin extends AutoPlugin {
-      Seq("ssh", "-i", sshKeyfile, "-l", sshUser)
+    val deploys = deployConfig.hostConfigs map { hostConfig =>
+      Future {
+        // Command to pass to rsync's "rsh" flag, and to use as the base of our ssh
+        // operations.
+        val sshCommand = {
+          val sshKeyfile = sys.env("AWS_PEM_FILE")
+          val sshUser = hostConfig.sshUser
+          Seq("ssh", "-i", sshKeyfile, "-l", sshUser)
+        }
+        val deployHost = hostConfig.host
+        val deployDirectory = hostConfig.directory
+
+        val rsyncDirs = deployDirs.value map (name => s"--include=/$name")
+        val rsyncCommand = Seq.concat(
+          Seq("rsync", "-vcrtzP", "--rsh=" + sshCommand.mkString(" ")),
+          rsyncDirs,
+          Seq(
+            "--exclude=/*",
+            "--delete",
+            universalStagingDir.getPath + "/",
+            deployHost + ":" + deployDirectory
+          )
+        )
+
+        // Shell-friendly version of rsync command, with rsh value quoted.
+        val quotedRsync = rsyncCommand.patch(
+          2, Seq("--rsh=" + sshCommand.mkString("\"", " ", "\"")), 1
+        ).mkString(" ")
+        log.info("Running " + quotedRsync + " . . .")
+        if (Process(rsyncCommand).! != 0) {
+          throw new IllegalArgumentException(s"Error running rsync to host '$deployHost'.")
+        }
+
+        // Now, ssh to the remote host and run the restart script.
+        val restartScript = deployDirectory + "/" + hostConfig.startupScript
+        val restartCommand = Seq.concat(
+          sshCommand,
+          Seq(deployHost, restartScript, "restart"),
+          hostConfig.startupArgs map { "--" +: _ } getOrElse Seq()
+        )
+        log.info("Running " + restartCommand.mkString(" ") + " . . .")
+        if (Process(restartCommand).! != 0) {
+          throw new IllegalArgumentException(s"Error running restart command on host '$deployHost'.")
+        }
+      } map Success.apply recover { case e => Failure(e) } // Map to Try so we can fold over errors.
     }
-    val deployHost = configMap("deploy.host")
-    val deployDirectory = configMap("deploy.directory")
-
-    val rsyncDirs = deployDirs.value map (name => s"--include=/${name}")
-    val rsyncCommand = Seq("rsync", "-vcrtzP", "--rsh=" + sshCommand.mkString(" ")) ++ rsyncDirs ++
-      Seq(
-        "--exclude=/*",
-        "--delete",
-        universalStagingDir.getPath + "/",
-        deployHost + ":" + deployDirectory
-      )
-
-    // Shell-friendly version of rsync command, with rsh value quoted.
-    val quotedRsync = rsyncCommand.patch(
-      2, Seq("--rsh=" + sshCommand.mkString("\"", " ", "\"")), 1
-    ).mkString(" ")
-    log.info("Running " + quotedRsync + " . . .")
-    if (Process(rsyncCommand).! != 0) {
-      throw new IllegalArgumentException("Error running rsync.")
+
+    // Accumulate all error messages thrown when trying to deploy.
+    val accumulatedErrors = Future.fold[Try[Unit], Option[String]](deploys)(None) {
```

Comment:
This is a bit confusing.

Also, I'd accumulate the full list of exceptions, then build a string later:
```scala
val errors = Future.fold[Try[Unit], Seq[Throwable]](deploys)(Seq.empty) {
  case (acc, Success(_)) => acc
  case (acc, Failure(e)) => acc :+ e
}
```